File: app/src/main/java/com/caglar/pokequery/domain/engine/GoalStringBuilder.kt
package com.caglar.pokequery.domain.engine

import com.caglar.pokequery.data.model.GeneratedString

/**
 * Builds the final generated string shown on a goal-detail screen from an engine
 * base goal plus the user's optional protection toggles.
 *
 * The base goal's `rawSyntax` is already language-translated by [StringBuilderEngine].
 * Optional protections are translated through the same [SearchTermMapper] before
 * deduplication so non-English outputs do not receive duplicate exclusions.
 */
object GoalStringBuilder {

    private val passthroughGoals = setOf("hundo_check", "nundo_finder", "pvp_candidates")

    private fun syntaxTokens(raw: String): Set<String> =
        raw.split('&', ',', ';', ':')
            .map(String::trim)
            .filter(String::isNotBlank)
            .toSet()

    fun buildFinal(
        baseGoal: GeneratedString,
        optionalProtections: List<String>,
        language: String = "English"
    ): GeneratedString {
        if (baseGoal.goalId in passthroughGoals) return baseGoal

        val existing = baseGoal.rawSyntax
        val existingTokens = syntaxTokens(existing)
        val translatedProtections = optionalProtections
            .distinct()
            .map { token -> "!${SearchTermMapper.translateSyntax(token, language)}" }

        val toAdd = translatedProtections
            .filterNot { it in existingTokens }
            .joinToString("&")

        if (toAdd.isEmpty()) return baseGoal

        val merged = if (existing.isBlank()) toAdd else "$existing&$toAdd"
        return baseGoal.copy(rawSyntax = merged)
    }
}

File: web/src/engine/goalStringBuilder.ts
// Port of com.caglar.pokequery.domain.engine.GoalStringBuilder

import type { GeneratedString } from '../types'
import { translateSyntax } from './searchTermMapper'

const passthroughGoals = new Set(['hundo_check', 'nundo_finder', 'pvp_candidates'])

function syntaxTokens(raw: string): Set<string> {
  return new Set(
    raw.split(/[&,;:]/)
      .map(token => token.trim())
      .filter(Boolean),
  )
}

export function buildFinal(
  baseGoal: GeneratedString,
  optionalProtections: string[],
  language: string = 'English',
): GeneratedString {
  if (passthroughGoals.has(baseGoal.goalId)) return baseGoal

  const existing = baseGoal.rawSyntax
  const existingTokens = syntaxTokens(existing)
  const translatedProtections = [...new Set(optionalProtections)]
    .map(token => `!${translateSyntax(token, language)}`)

  const toAdd = translatedProtections
    .filter(token => !existingTokens.has(token))
    .join('&')

  if (!toAdd) return baseGoal

  const merged = existing ? `${existing}&${toAdd}` : toAdd
  return { ...baseGoal, rawSyntax: merged }
}
